### PR TITLE
Add xtree

### PR DESCRIPTION
**Pull request by isbm**
Adds image appliance description inheritance.
- [x] Add
- [x] Merge
- [x] Update
- [x] Remove
- [x] Set attributes (YAML)
- [x] XML appliance Loader
- [x] Unit tests

XTree allows you to inherit Appliance descriptions and build a chain of these in any way you want. Usage:

```python
from berry_mill.imgdescr.loader import Loader
final_rendered_xml_string = Loader.load("/path/to/the/appliance.xml")

# Save it somewhere else for the future work
with open("/tmp/my_merged_appliance.xml", "w") as wfd:
    wfd.write(final_rendered_xml_string)
```

The content of the loaded `appliance.xml` XML above can be as simple as that:

```xml
<?xml version="1.0" encoding="utf-8"?>
<image schemaversion="6.8" name="Ubuntu-22.04_appliance">
    <inherit path="path/to/base-appliance.xml"/>
</image>
```

In this case it will do nothing to the `base-appliance.xml`, retaining its content as is. Appliance can also inherit other inherited appliance and build an infinite path of these.

Multiple inheritance is not supported.

### CODE REVIEW

File: test/test_traverse.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+
+        assert l._Loader__i_stack == ["test/test_appliance.xml"], "Wrong root traversal"
+
+    def test_loader_traversal_flatten_c(self):
+        """
+        Load full chain, changes from the level C should be applied.
+        """
+        l:Loader = Loader()
+        lusers = ApplianceDescription.find_all("user", ET.fromstring(l.load("test/chain_d.xml")))
+
```
Comment (wintron04) [inline test/test_traverse.py:55]: `assert "pwdformat" in luser[0].attrib, "Password in Plain, pwdformat can't be default"`
`assert luser[0].attrib["pwdformat"] == "plain", "Password format mismatch"`

File: test/appliance_add_packages.xml [hunk 2]
```diff
@@ -0,0 +1,72 @@
+<?xml version="1.0" encoding="utf-8"?>
+
+<!-- Assuming "test_appliance.xml" is in the same directory -->
+<image schemaversion="6.8" name="Ubuntu-22.04_appliance">
+    <inherit path="test/test_appliance.xml"/>
```
Comment (wintron04) [inline test/appliance_add_packages.xml]: Usage: `berrymill -i /path/to/appliance.xml`

berrymill will cd in dir "/path/to"

Also with the change I proposed to the Loader.load() method this only works for me with:
`<inherit path="test_appliance.xml"/>`
(Assuming ... is in the same dir)

**Review by wintron04 — COMMENTED** [review]
Suggestions

File: src/berry_mill/imgdescr/descr.py [hunk 3]
```diff
@@ -1,2 +1,234 @@
...
+    __P_AD = "add"
+    __P_RM = "remove"
+    __P_MG = "merge"
+    __P_RP = "replace"
+    __P_RA = "remove_any"
+    __P_ST = "set"
+
+    def __init__(self, descr: str, parent: str = None) -> ApplianceDescription:
+        self.s_dom: ET.Element = ET.fromstring(descr.encode("utf-8"))
+        self.p_dom: ET.Element|None = parent and ET.fromstring(parent.encode("utf-8")) or None
```
[the remark this answers is not in the document]
Comment (isbm) [reply src/berry_mill/imgdescr/descr.py]: Yes, better use ternary. I updated this.

File: src/berry_mill/imgdescr/loader.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        with open(next(iter(self.__i_stack))) as fd:
+            descr = ApplianceDescription(fd.read())
+
+        for pth in self.__i_stack[1:]:
+            with open(pth) as fp:
+                descr = ApplianceDescription(fp.read(), descr.to_str())
+
+        return descr.to_str()
+
+    def load(self, pth: str) -> str:
```
[the remark this answers is not in the document]
Comment (isbm) [reply src/berry_mill/imgdescr/loader.py:54]: But that has nothing to do with it. User defines the tree. Best if user does something like `path="../base.xml"` or `path="base.xml"` etc. Which means you first render the tree, then grab the whole output as _one single XML file_, dump it elsewhere, copypaste all root/cfg stuff, cd there, and then build the appliance as usual Kiwi does.

[the remark this answers is not in the document]
Comment (isbm) [reply src/berry_mill/imgdescr/loader.py:54]: Surely, you need to render the tree from the correct directory point, hence you `cd` to that inherited appliance _first_, and then render, and then move elsewhere you need... That is the only limitation. But that is the same as `Makefile` would work or any compiler: as long as it finds the root src, it will then resolve it. That part of the integration is outside of the scope of Xtree class.

File: test/appliance_add_packages.xml (hunk 2 above)
Comment (wintron04) [reply test/appliance_add_packages.xml]: What I mean is:
appliance_add_packages.xml makes the Assumption that test_appliance.xml is in the same dir.
The test_appliance you added is (for testing) most likely the base appliance (I am assuming)

When the User invokes berrymill they need to add the path to the appliance.xml (appliance_add_packages.xml)

The current behavior of berrymill is to change dir into that directory. 
And with that your _traverse() function reads path="test/test_appliance.xml", tries to find that and cant because its really not inside another test dir form that perspective.

Long story short:
I dont know why it works in the unit test but it doesnt work in "production"
we need to think of a coherent working directory structure. 
If you know what I mean

**Review by wintron04 — APPROVED** [review]
LGTM!
